---
title: "TF_Motif_Mapping"
author: "Matt Richards"
date: "6/26/2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Problem Intro

At this point, we have a couple of really useful assets:

* A collection of databases that map footprints to motifs for different tissues.
* A set of motifs from MotifDb that constitute the full set of motifs used for the databases.
* An R package (TReNA) containing tools that aim to use the footprint databases to narrow down the predictor space of a regression problem.

So if we're looking at this more systematically, it's something like:

Motifs $\rightarrow$ Databases $\rightarrow$ TReNA 

But what are these arrows in reality; or, rephrased, what tools or concepts connect our assets together? The first arrow is chronicled in the BDDS repository as a couple of workflows that take in motifs and footprints, then output our databases. The first workflow, chronicled [here](https://github.com/PriceLab/BDDS/tree/master/trena/fimo), takes motifs and uses the FIMO method of the [MEME suite]() to create a "FIMO" database of all motifs found in the human genome. The second workflow, chronicled [here](https://github.com/PriceLab/BDDS/tree/master/footprints/testdb), takes the "FIMO" database plus some tissue-specific footprints and uses them to create a tissue-, method-, and seed-specific database. As an example of this second workflow, we may have a set of footprints for brain tissue the we created using the HINT method for seed 20. Our method would intersect these footprints with FIMO to create a database called "brain_hint_20". 

All of this is well and good, but now we move to the second "arrow" of our paradigm; that is, how do we go from databases to TReNA? TReNA deals not with motifs, but with transcription factors (TFs), and our databases essentially put footprints in the context of motifs. What we need is a good motif-TF mapping, and at the moment we don't have one. But the intention of this document is to arrive at a definitive mapping by the end. 

## Inspecting Existing Info

We don't have a definitive mapping yet, otherwise we wouldn't be undergoing this exercise; but what, in fact, DO we have? Here's another list of assets, this one pertaining specifically to our motif-TF mapping:

1. Our final list of motifs from MotifDb (all Jaspar2016 or HOCOMOCOv10)
2. Metadata for all our motifs, contained within MotifDb
3. Raw Jaspar-formatted files for the motifs themselves in terms of PCM
4. A rigorously documented process through which we arrived at our final motif list
5. TOMTOM files that compare HOCOMOCO motifs with Jaspar to find motifs that align
6. The full set of human/mouse motifs in both Jaspar and HOCOMOCO from MotifDb

All of these are potentially important; asset #1 is the full dataset we want to be able to map to/from, so that's obviously important. The trick is that the motifs in this list aren't really the only set of motifs we care about because we ran TOMTOM (assets #4-5) to reduce redundancy in the full list of motifs (asset #6). So although we ultimately want to map just the motifs from asset #1, we want to do so in a way that accounts for all motifs in asset #6, and the record of how we do this is found in assets #4-5. It may be that these 4 assets (or maybe just 3 of them) will prove sufficient for our purposes, but we also have asset #2, which specifically gives TF names for motifs. And of course, we could go a level higher to asset #3 and check the source material for MotifDb if need be; however, it's pretty unlikely we'll need to do this. 

Based upon this quick little overview of different data bits and pieces, here's (another enumerated) plan:

1. Load MotifDb to give us access to all of its metadata (asset #2)
2. Load the list of motifs used for our final FIMO (asset #1)
3. Load the list of all human/mouse motifs from HUMAN/HOCOMOCO (asset #6)
4. Load the TOMTOM files of comparisons (asset #5)
5. Using assets #2 and #6 to create a mapping of all motifs to all TFs (asset #7)
6. Using assets #1 and #5, create a mapping of all "redundant" motifs to their "parent" motif (asset #8)
7. Using our two new assets (#7-8), create a mapping of all "parent" motifs to all TFs

Looks like a plan; let's dive right in and load up our datasets

## Steps 1-4: Loading all the files

We'll load all the different files and libraries; first we'll load MotifDb and the dplyr library for our inevitable data manipulation.

```{r message = FALSE}
library(MotifDb); library(dplyr)
```

We can get our set of all human and mouse motifs from Jaspar2016 and HOCOMOCOv10 using a simple query or two:

```{r}
all.motifs <- subset(MotifDb, dataSource %in% c("HOCOMOCOv10","jaspar2016"))
all.motifs <- subset(all.motifs, organism %in% c("Hsapiens","Mmusculus"))
length(all.motifs)
```

That's assets #2 and #6; in order to get the other 2 assets, we'll borrow some code from asset #4. First we'll grab asset #5 by reading the TOMTOM files and sticking them together:

```{r}
human.hoco <- read.table(file = "./TOMTOM/hocomoco_jaspar_human/tomtom.txt")
mouse.hoco <- read.table(file = "./TOMTOM/hocomoco_jaspar_mouse/tomtom.txt")
all.hoco <- rbind(human.hoco,mouse.hoco)
names(all.hoco) <- c("Query.ID", "Target.ID", "Optimal.offset", "Pvalue","Evalue","Qvalue","Overlap","Query.consensus","Target.consensus","Orientation")
dim(all.hoco)
```
This matches what we had in asset #4; we'll go a step further and trim off everything but what we need (motifs and Pvalue), plus add a logP value to match our other analysis

```{r}
all.hoco <- all.hoco %>% 
  select(Query.ID, Target.ID, Pvalue) %>%
  mutate(logP = -1*log10(Pvalue))
head(all.hoco)
```

Looks good; now then, we want to follow the process from asset #4 to get our list of final motifs. This means screening out anything with a logP value below 7.3. We'll also change the Query.ID and Target.ID columns to characters, just for convenience later: 

```{r}
above.p <- all.hoco %>% filter(logP >= 7.3)
above.p$Query.ID <- as.character(above.p$Query.ID)
above.p$Target.ID <- as.character(above.p$Target.ID)
repeat.motifs <- unique(above.p$Query.ID)
length(repeat.motifs)
```
This matches what we expect; 417 HOCOMOCO motifs qualify as repeats and can be removed. We're going slowly, but that's because we're just being extra careful as we go along to make sure everything looks good. We'll now go through the process of removing these repeats from our full motifs list, which should leave us with 1256 motifs instead of the 1673 we had before:

```{r}
trimmed.repeats <- gsub(".+HOCOMOCOv10-","",repeat.motifs)
fimo.motifs <- subset(all.motifs, !(providerId %in% trimmed.repeats))
length(fimo.motifs)
```

Looks great, so let's review what we have before taking the next steps:

1. Asset #2: MotifDb, the set of all motifs and metadata (in the MotifDb variable)
2. Asset #6: Our list of all Jaspar and HOCOMOCO motifs (in the all.motifs variable)
3. Asset #5: The TOMTOM record that matches motifs together (in the above.p variable)
4. Asset #1: Our list of final FIMO motifs (in the fimo.motifs variable)

Now we'll put these together to get our final list. 

## Steps 5-7: Map assets to create final list

To remind ourselves of the first of these steps, here it is:

5. Using assets #2 and #6 to create a mapping of all motifs to all TFs (asset #7)

We don't actually need to mess with the MotifDb variable to do this, we just needed to load the library. Our other asset is a MotifDb object, so it's already got the metadata we need. Let's see about creating our mapping. First, where is our data? We can access values using the `value` command; MotifDb has a structure of values as follows:

```{r}
colnames (values (all.motifs))
```

What we actually want here is a combination of the "providerID" and "geneSymbol" data:

```{r}
head(values(all.motifs)$providerName)
head(values(all.motifs)$geneSymbol)
```
As demonstrated by these top entries, the TFs match up to their motifs. So we'll make a data frame using these 2 pieces of info and our job is done for step 5.  

```{r}
asset.7 <- data.frame(motifs = as.character(values(all.motifs)$providerName),
                      TFs = as.character(values(all.motifs)$geneSymbol))
head(asset.7); dim(asset.7)
```

We'll also do the same for the fimo motifs; this is going to come in handy later:

```{r}
fimo.TFs <- data.frame(motifs = values(fimo.motifs)$providerName,
                      TFs = values(fimo.motifs)$geneSymbol)
head(fimo.TFs); dim(fimo.TFs)
```

Moving on, we now come to step 6:

6. Using assets #1 and #5, create a mapping of all "redundant" motifs to their "parent" motif (asset #8)

So essentially, what we want is another data frame with 2 columns; the first will be a "child" motif from HOCOMOCO and the second will be a "parent" motif from Jaspar that matches it. These won't be 1:1 or 1:n matches, they'll be n:m, so we'll have to make sure to do this cleverly. Let's take a look at the top of the TOMTOM data frame so we get a better idea of what we're working with:

```{r}
head(above.p); dim(above.p)
```

One thing to notice is that our motif names are a little different; here, they're preceeded by the species and data source. We'll want to trim those or append our MotifDb motif names in order to make them work for us. Let's go ahead and trim them all, plus get rid of the columns we no longer need. It's worth noting that the Jaspar motifs actually trim everything before the "MA..." bit, so make sure to do that too:

```{r}
above.p$Query.ID <- gsub(".+HOCOMOCOv10-","",above.p$Query.ID)
above.p$Target.ID <- gsub(".+jaspar2016-.+-","",above.p$Target.ID)
asset.8 <- select(above.p, Query.ID, Target.ID); head(asset.8)
```

This actually gives us exactly what we wanted, so we're done with this step and we'll move on to step 7:

7. Using our two new assets (#7-8), create a mapping of all "parent" motifs to all TFs

This is a bit trickier; the first thing we'll do is a left join that maps all our "Query" motifs to their transcription factors. 

```{r message = FALSE}
prelim.mapping <- left_join(asset.8,asset.7, by = c("Query.ID" = "motifs"))
head(prelim.mapping);dim(prelim.mapping)
```
Now we've got a mapping of our "Target" motifs to the TFs, so we can put this together with our initial list of TFs, our `fimo.TFs`. This may result in some redundancy, but we'll take care of that in a minute. We'll also change a column name while we're at it:

```{r}
names(prelim.mapping)[2] <- "motifs"
final.mapping <- rbind(prelim.mapping[,2:3], fimo.TFs)
head(final.mapping);dim(final.mapping)
```

There's undoubtedly some duplicated rows here, so let's screen those out and call it finished

```{r}
final.mapping <- final.mapping[!duplicated(final.mapping),]
dim(final.mapping); head(final.mapping)
```

As a final sanity check, we should make sure there are still only 1256 unique motif values as well:

```{r}
length(unique(final.mapping$motifs))
```

## Inspecting the Mapping

Let's look at how the motifs and TFs map:

```{r message = FALSE}
library(ggplot2)
final.mapping %>% 
  group_by(motifs) %>% 
  summarize(TFs = n()) %>% 
  arrange(desc(TFs)) %>% 
  ggplot(aes(x=TFs)) + geom_histogram(bins = 12)
```

```{r}
final.mapping %>% 
  group_by(TFs) %>% 
  summarize(motifs = n()) %>% 
  arrange(desc(motifs)) %>% 
  ggplot(aes(x=motifs)) + geom_histogram(bins = 17)
```

We'll also compare to Seth's previous mapping:

```{r}
seth.table <- read.table(system.file(package="trena", "extdata", "motifGenes.tsv"), sep = "\t", header=TRUE)
dim(seth.table)
```

We can see that Seth's original mapping is much more extensive in terms of total TF-motif relationships. However, it doesn't actually cover nearly as many motifs:

```{r}
length(unique(seth.table$motif))
```
So what we essentially have here are 2 very different mappings. Our previous mapping covered fewer motifs and TFs, but was very well connected, with some motifs being connected to dozens of different TFs and visa versa. Our new mapping covers many more motifs, more than twice as many, but the mapping is dominated by 1-1 connections, with only a vast minority of motifs exceeding a handful of TFs. We can see this by looking at the plots of Seth's table

```{r}
seth.table %>% 
  group_by(motif) %>% 
  summarize(Num.Genes = n()) %>% 
  arrange(desc(Num.Genes)) %>% 
  ggplot(aes(x=Num.Genes)) + geom_histogram(bins = 50)
```

```{r}
seth.table %>% 
  group_by(tf.gene) %>% 
  summarize(motif = n()) %>% 
  arrange(desc(motif)) %>% 
  ggplot(aes(x=motif)) + geom_histogram(bins = 50)
```

Of particular concern is whether we're missing any major things from Seth's data table; what are the things missing from our table but present in his (for motifs)? 

```{r}
missing <- setdiff(seth.table$motif,final.mapping$motifs)
length(missing)
```

As we can see, there are 81 motifs missing from our mapping; now a lot of these (actually, 67 of 81) are from other databases that we're not going to deal with. But 14 of them begin with "MA...", indicating they're from JASPAR. Why didn't we grab these? Because they're from other vertebrates. Here's the list of them:

```{r}
missing.jaspar <- grep("^MA\\d+",missing, value = TRUE)
missing.jaspar
```

If we like, we can grab these entries from MotifDb and look at exactly where they're from:

```{r}
jaspar.metadata <- subset(query(MotifDb, "jaspar2016"), providerId %in% missing.jaspar)
table(values(jaspar.metadata)$organism)
```

Looks like 8 are from 2 different rat species, 3 are from a rooster, and 1 is from a rabbit. Now we could just add these directly, but it's much more rigorous if we instead add all motifs from vertebrates that we haven't yet included. Here's the list of all species in jaspar2016:

```{r}
table(values(query(MotifDb, "jaspar2016"))$organism)
```

With a little bit of checking around, we can make a list of the non-Human and non-Mouse vertebrates in this list:

```{r}
vertebrates <- c("Ggallus","Nsp.","Ocuniculus","Rnorvegicus","Rrattus","Vertebrata", "Xlaevis",NA)
new.motifs <- subset(query(MotifDb, "jaspar2016"), organism %in% vertebrates)
length(new.motifs)
values(new.motifs)$providerId
```

Now with our list of new motifs, we should cover all of Seth's motifs, but let's see if we do:

```{r}
setdiff(missing.jaspar, values(new.motifs)$providerId)
```

What are these 2 motifs? Let's investigate further:

```{r}
last.two <- setdiff(missing.jaspar, values(new.motifs)$providerId)
last.two <- subset(query(MotifDb, "jaspar2016"), providerId %in% last.two)
values(last.two)$organism
```
Looks like these 2 are in MotifDb, but neither of them has an organism associated with them in the Jaspar metadata. Something to keep in mind; let's move on to looking at transcription factors for now. There's a disparity in what was covered by Seth's table and what TFs we covered in the "final.mapping":

```{r}
length(unique(seth.table$tf.gene))
length(unique(final.mapping$TFs))
length(setdiff(seth.table$tf.gene,final.mapping$TFs))
```

Unfortunately, we're missing 329 of the TFs from Seth's mapping, which definitely isn't ideal. Now let's ask the following question: how many of those TFs are associated with:

A. The 81 motifs we don't yet have
B. The 24 motifs we COULD add
C. The 67 motifs we don't intend to add

First, here's how many we're missing because of 81; this is the number that, even if we put the tables together, won't be covered because of the missing motifs:

```{r}
# Get motifs from 81
missing.tfs <- setdiff(seth.table$tf.gene,final.mapping$TFs)
only.from.81 <- seth.table %>% filter(motif %in% missing)
tfs.from.81 <- unique(only.from.81$tf.gene)

# Get motifs not from 81
not.from.81 <- seth.table %>% filter(!(motif %in% missing))
tfs.not.81 <- unique(not.from.81$tf.gene)

# Find the difference in the 2
tfs.only.in.81 <- setdiff(tfs.from.81, tfs.not.81)
# Total number of TFs we're missing
print(sprintf("Total number of missing TFs: %d",length(missing.tfs)))
# Number of TFs we're missing that don't have to do with the 81
print(sprintf("Missing TFs mapped to motifs we have: %d",length(setdiff(missing.tfs, tfs.only.in.81))))
# Number of Tfs associated with the 81 that are not missing
print(sprintf("Existing TFs tied to the 81 missing motifs: %d", length(setdiff(tfs.only.in.81, missing.tfs))))
# Numer of TFs we NEED to get from the 81
print(sprintf("Number of TFs we can only get from the 81 missing motifs: %d",length(intersect(tfs.only.in.81,missing.tfs))))

```

What this tells us is that even if we add no more motifs, Seth's table is going to add 247 TFs to motifs we already have. The missing 81 motifs map to 21 TFs we've already got in our table and 82 that we don't. So it's those 82 we'd be missing if we just smushed the mappings from our table and Seth's table together for our motif list. Now then, let's see how many of these TFs we could potentially add from our 24 motifs:

```{r}
# Grab the 82 TFs that are still missing
still.missing <- intersect(tfs.only.in.81,missing.tfs)

# Grab the TFs associated with the 24 new genes from JASPAR
from.jaspar <- values(new.motifs)$geneSymbol

# Grab the TFs associated with the 24 new genes from Seth's table
new.motif.names <- values(new.motifs)$providerId
seth.tfs <- seth.table %>% filter(motif %in% new.motif.names)
from.seths <- as.character(seth.tfs$tf.gene)

# Add the 2 sets, find number in common with the 103 missing
all.new.tfs <- c(from.jaspar,from.seths)
intersect(all.new.tfs,still.missing)
```
Looks like we'd only be adding 2 more TFs. Now 2 motifs from Jaspar still exist outside of our 24, but let's see how many genes are associated with the other 67 TFs:

```{r}
# Grab the names of the 67 other motifs
missing.jaspar.idx <- grepl("^MA\\d+",missing)
non.jaspar <- missing[!missing.jaspar.idx]

# Find only the TFs associated with these motifs
seth.tfs <- seth.table %>% filter(motif %in% non.jaspar)
non.jaspar.tfs <- seth.tfs$tf.gene

# Find the intersection with the 101
all.new.tfs <- intersect(all.new.tfs,still.missing)
final.80 <- setdiff(still.missing, all.new.tfs)
length(intersect(non.jaspar.tfs, final.80))
```
Looks like we get 78 of the remaining 80 TFs from these 67 non-JASPAR motifs. So we could infer from this that the 2 missing JASPAR motifs give us the final 2, but let's check to make sure:

```{r}
final.2 <- setdiff(final.80, non.jaspar.tfs)
jaspar.2 <- setdiff(missing.jaspar, values(new.motifs)$providerId)
jaspar.2.tbl <- seth.table %>% filter(motif %in% jaspar.2)
jaspar.2.tfs <- jaspar.2.tbl$tf.gene

# Find the intersection
intersect(final.2, jaspar.2.tfs)
```

So it seems we can do a handful of things:

1. Ignore extra motifs; we'll still cover 247 of the 329 TFs, even if we add nothing
2. Add the 24 JASPAR motifs from vertebrates plus "NA", giving 58 more motifs than we had before, but only 4 more TFs to leave 78 more (note, we could add just the 24, which would give us 2 more TFs too)
3. Add the aforementioned JASPAR motifs, plus all the other 67 motifs from Jolma/Uniprot/other that Seth had, but nothing else
4. Add the JASPAR motifs and then find resources with other motifs that include the other 67, but have a logical method for picking them. 

Just as a sample, if we don't add anything past more JASPAR motifs, here's the list of TFs (78) that we'll miss, in alphabetical order:

```{r}
sort(intersect(non.jaspar.tfs, final.80))
```

If we don't add anything, we'd also miss these 4:

```{r}
intersect(all.new.tfs,still.missing)
```

Also of interest: here are the non-Jaspar motifs we'd add (the 67):

```{r}
non.jaspar
```
So...what now? We need some way to deal with these missing motifs. To that end, we'll continue this in another document, `swiss_uniprobe_compare.Rmd`. For that document, we'll go ahead and save our `final.mapping` variable. 

```{r}
saveRDS(final.mapping,"./jaspar_hocomoco_mapping.RDS")
```

So ends our initial mapping document. We've got a first mapping, but we need to add a good 82 TFs, most of which will come from other databases. 